### PR TITLE
build-init plugin and --incubating option 2

### LINKED ISSUE #18238: Update build init to provide opt-in to best practices based on incubating APIs

### PR DESCRIPTION
**Pull request by tresat**
Fixes #18238 

sanityCheck, build-init:quickTest, docs:serveDocs working, with the exception of 2 kotlin library build init integration tests here: https://ge.gradle.org/s/7g7rw43dxzhpi/tests/overview?outcome=failed (and Swift tests, which fail for me locally).

### CODE REVIEW

**Comment by big-guy** [comment]
@bot-gradle test QF

File: subprojects/docs/src/samples/build-organization/incubating/publishing-convention-plugins/kotlin/convention-plugins/src/main/kotlin/com.myorg.service-conventions.gradle.kts
Comment (big-guy) [inline subprojects/docs/src/samples/build-organization/incubating/publishing-convention-plugins/kotlin/convention-plugins/src/main/kotlin/com.myorg.service-conventions.gradle.kts:20]: we should be able to switch to registering when the other PR I mentioned above is merged

File: subprojects/docs/src/samples/java/incubating/modules-multi-project-with-integration-tests/kotlin/buildSrc/src/main/kotlin/myproject.java-conventions.gradle.kts
Comment (big-guy) [inline subprojects/docs/src/samples/java/incubating/modules-multi-project-with-integration-tests/kotlin/buildSrc/src/main/kotlin/myproject.java-conventions.gradle.kts:15]: With https://github.com/gradle/gradle/pull/18308, we'll be able to drop the dependencies block below and just use `useJUnitJupiter("5.7.1")`

**Review by big-guy — CHANGES_REQUESTED** [review]
I've gotten through about 3/4 of the files changed. I need to look harder at the build script builder changes and the samples, but I hope my first round of comments are helpful.

I think this is going in the right direction.

File: subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/AbstractInitIntegrationSpec.groovy
Comment (tresat) [reply subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/AbstractInitIntegrationSpec.groovy:63]: It was a little more complicated, because of how some of the `TestExecutionResult` subclasses made assumptions that a least one test would have been run, so I just did it this way to get something working quickly.  I've replaced with the proper methods for each implementation.

File: subprojects/docs/src/samples/build-organization/incubating/publishing-convention-plugins/groovy/convention-plugins/build.gradle
Comment (tresat) [reply subprojects/docs/src/samples/build-organization/incubating/publishing-convention-plugins/groovy/convention-plugins/build.gradle:37]: Had to remove the dep on the BOM, since the platform call wasn't working and I didn't want to try to add it to ComponentDependencies.

File: subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/JavaApplicationInitIntegrationTest.groovy
Comment (tresat) [reply subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/JavaApplicationInitIntegrationTest.groovy]: That fixed it!

**Comment by big-guy** [comment]
@bot-gradle test this

File: subprojects/docs/src/samples/build-organization/incubating/publishing-convention-plugins/groovy/convention-plugins/src/main/groovy/com.myorg.service-conventions.gradle
Comment (big-guy) [inline subprojects/docs/src/samples/build-organization/incubating/publishing-convention-plugins/groovy/convention-plugins/src/main/groovy/com.myorg.service-conventions.gradle:17]: I think this can be removed now

File: subprojects/docs/src/samples/build-organization/incubating/publishing-convention-plugins/groovy/convention-plugins/build.gradle
Comment (big-guy) [inline subprojects/docs/src/samples/build-organization/incubating/publishing-convention-plugins/groovy/convention-plugins/build.gradle:33]: This isn't really correct because this causes us to add JUnit Jupiter dependencies as well as the Spock dependencies below. 

I think it's probably easiest for us to go ahead and add a `useSpock(...)` convenience method. Let's leave this as-is and do that in a separate change.

Alternatively, we could remove `useJUnitJupiter` and configure `useJUnitPlatform` directly on the task.

File: subprojects/docs/src/samples/build-organization/incubating/publishing-convention-plugins/kotlin/convention-plugins/src/main/kotlin/com.myorg.service-conventions.gradle.kts
Comment (big-guy) [inline subprojects/docs/src/samples/build-organization/incubating/publishing-convention-plugins/kotlin/convention-plugins/src/main/kotlin/com.myorg.service-conventions.gradle.kts:13]: Is this different from the Groovy version?

File: subprojects/docs/src/samples/java/incubating/jvm-multi-project-with-additional-test-types/groovy/buildSrc/src/main/groovy/myproject.java-conventions.gradle
Comment (big-guy) [inline subprojects/docs/src/samples/java/incubating/jvm-multi-project-with-additional-test-types/groovy/buildSrc/src/main/groovy/myproject.java-conventions.gradle:18]: I think this can be removed now

File: subprojects/docs/src/samples/java/incubating/jvm-multi-project-with-additional-test-types/kotlin/buildSrc/src/main/kotlin/myproject.java-conventions.gradle.kts
Comment (big-guy) [inline subprojects/docs/src/samples/java/incubating/jvm-multi-project-with-additional-test-types/kotlin/buildSrc/src/main/kotlin/myproject.java-conventions.gradle.kts:15]: This is different from the groovy version?

File: subprojects/docs/src/samples/java/incubating/jvm-multi-project-with-additional-test-types/kotlin/application/build.gradle.kts
Comment (big-guy) [inline subprojects/docs/src/samples/java/incubating/jvm-multi-project-with-additional-test-types/kotlin/application/build.gradle.kts:8]: I think this is the same sample project as with the other multi-project applications. Do you know why this sometimes has an implementation dependency on utilities AND list and sometimes it's just utilities? e.g., see the modules-multi-project-with-integration-tests sample below.

File: subprojects/docs/src/samples/java/incubating/modules-multi-project-with-integration-tests/groovy/buildSrc/src/main/groovy/myproject.java-conventions.gradle
Comment (big-guy) [inline subprojects/docs/src/samples/java/incubating/modules-multi-project-with-integration-tests/groovy/buildSrc/src/main/groovy/myproject.java-conventions.gradle:18]: I think this can be removed

File: subprojects/docs/src/samples/java/incubating/modules-multi-project-with-integration-tests/kotlin/buildSrc/src/main/kotlin/myproject.java-conventions.gradle.kts
Comment (big-guy) [inline subprojects/docs/src/samples/java/incubating/modules-multi-project-with-integration-tests/kotlin/buildSrc/src/main/kotlin/myproject.java-conventions.gradle.kts:15]: This is different from the Groovy version

**Review by big-guy — CHANGES_REQUESTED** [review]
Some comments about the samples.  I still need to look more at BuildScriptBuilder, JvmProjectInitDescriptor and JvmGradlePluginProjectInitDescriptor, but everything else seems OK apart from my comments.

File: subprojects/docs/build.gradle
Comment (tresat) [reply subprojects/docs/build.gradle:236]: I'll put them all into a separate category.

File: subprojects/docs/src/samples/java/incubating/jvm-multi-project-with-additional-test-types/kotlin/application/build.gradle.kts
Comment (tresat) [reply subprojects/docs/src/samples/java/incubating/jvm-multi-project-with-additional-test-types/kotlin/application/build.gradle.kts]: I think this is because in this project the `"utilities` project had only an `implementation` dependency on `:list` until we updated it on 8/19.  I'll get these to only dep `:utilities`.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle cancel

**Comment by tresat** [comment]
@bot-gradle clean

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@DPUkyle, @big-guy  - All comments addressed.  This is ready to be re-reviewed.  It only covers Java Lib and Applications, and Kotlin Libs.  The other options I can do one at a time in separate PRs after this is merged in, I don't want to hold this PR up working on other project types.

File: subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/JavaApplicationInitIntegrationTest.groovy
Comment (DPUkyle) [inline subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/JavaApplicationInitIntegrationTest.groovy:47]: Maybe annotate with `@Unroll` and add a `#` variable name in the test name?

Since the two states you are testing are so similar, I might rewrite as:

```
def 'defaults to Groovy build scripts when incubating flag=#isIncubating'() {
  ...
  when:
  run(['init', '--type', 'java-application', isIncubating ? '--incubating' : null])
  ...
  where:
  isIncubating << [true, false]
}

```

File: subprojects/build-init/src/testFixtures/groovy/org/gradle/buildinit/plugins/fixtures/ScriptDslFixture.groovy
Comment (DPUkyle) [inline subprojects/build-init/src/testFixtures/groovy/org/gradle/buildinit/plugins/fixtures/ScriptDslFixture.groovy:151]: I know it's just a test fixture and not production code, but do we really want to be parsing scripts? I see maintenance headaches. Let's talk options.

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmTestSuite.java
Comment (DPUkyle) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmTestSuite.java:47]: Thank you for adding! You will likely have to rebase this after #18343 is merged; sorry!

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java
Comment (DPUkyle) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java:50]: nice!

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java
Comment (DPUkyle) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java:145]: @big-guy Doesn't spock require the vintage engine in order to use JUnit Jupiter? Or no since it is Spock 2.0-groovy-3.0?

**Review by DPUkyle — CHANGES_REQUESTED** [review]
Thanks Tom!

Some general themes-

- Please use `@Unroll` wherever possible for spock tests that you touched or added
- Check bean naming conventions for boolean properties
- A few tests were added but I didn't see how there are related to init-plugins
- Could the "frameworks" piece have been separated into another PR?

File: subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/InitSettings.java
Comment (tresat) [reply subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/InitSettings.java:120]: I think either is permitted for booleans, but I agree is sounds better here and will change.

File: subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java
Comment (tresat) [reply subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:120]: This one has to begin with get, see `getSplitProject()`.

File: subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java
Comment (tresat) [reply subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:240]: It is absent if not set with the argument.  getOrElse() is a good idea to condense this.

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java
Comment (tresat) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java:145]: Looks like the need for Vintage was removed with 2.0.

File: subprojects/internal-testing/src/main/groovy/org/gradle/integtests/fixtures/HtmlTestExecutionResult.groovy
Comment (tresat) [reply subprojects/internal-testing/src/main/groovy/org/gradle/integtests/fixtures/HtmlTestExecutionResult.groovy:78]: There are actually 4 implementations of `TestExecutionResult` which each need to override this a little differently.  I guess I moved away from using it in favor of `assertTestClassesNotExecuted` to be a little more lenient when debugging.  I could probably switch back now.  Even if both are not currently used, I think they're worth keeping as 2 separate types of checks in the fixture.

File: subprojects/docs/src/samples/incubating/build-organization/publishing-convention-plugins/kotlin/convention-plugins/src/test/kotlin/com/myorg/ServicePluginTest.kt
Comment (tresat) [reply subprojects/docs/src/samples/incubating/build-organization/publishing-convention-plugins/kotlin/convention-plugins/src/test/kotlin/com/myorg/ServicePluginTest.kt:10]: When updating the samples, I updated everything that had integrationTests defined.  The `publishing-convention-plugins` sample uses it in a convention plugin.

**Comment by big-guy** [comment]
> Thanks Tom!
> 
> Some general themes-
> 
> * Please use `@Unroll` wherever possible for spock tests that you touched or added
> * Check bean naming conventions for boolean properties
> * A few tests were added but I didn't see how there are related to init-plugins
> * Could the "frameworks" piece have been separated into another PR?

@DPUkyle Just FYI, Spock 2 should automatically unroll tests by default now, so adding `@Unroll` isn't necessary.

File: subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/JavaApplicationInitIntegrationTest.groovy
Comment (big-guy) [reply subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/JavaApplicationInitIntegrationTest.groovy]: Don't change it back, but IMHO the explicit list is easier to understand than a boolean in a test.

We could have done this to make it clearer what the test was doing:
```
def "defaults to Groovy build scripts #tasks"() {
```
A lot of the time, a boolean in a test screams "there are two different tests here" whereas two sets of arguments feel like the same test. But six of one, half a dozen of the other. The latest changes are OK.

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle cancel

**Comment by tresat** [comment]
@bot-gradle test this

**Review by big-guy — APPROVED** [review]
Let's get this merged and follow up on anything bad in a subsequent PR

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test and merge

File: subprojects/build-init/src/testFixtures/groovy/org/gradle/buildinit/plugins/fixtures/ScriptDslFixture.groovy
Comment (DPUkyle) [reply subprojects/build-init/src/testFixtures/groovy/org/gradle/buildinit/plugins/fixtures/ScriptDslFixture.groovy:151]: Tabling for now

**Review by DPUkyle — APPROVED** [review]

**Comment by blindpirate** [comment]
Apart from the Test Distribution failure, I guess the samples in docs need to be fixed, too: https://ge.gradle.org/s/pgkbvow4ttpsg

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Pull request merged**
